### DISCUSSION TITLE
Noob question: how does gluetun know which port belongs to which container??!

### DISCUSSION THREAD

**Discussion by Kamikaze01git** [qdm12/gluetun#1566] [Q&A] [answered]
Hello there :)

I have successfully running a gluetun container.

Additional I have two containers running with
`network_mode: container:gluetun`

And inside gluetun container I use
```dockerfile
ports:
   - 7181:7181 #container 1
   - 5031:5031 #container 2
``` 
So far so good: Everything works like a charm and I can reach container 1 with port 7181 and container 2 with port 5031. Both working over VPN 👍

**BUT:** because I only specify  `network_mode: container:glutun` without any ports in both of my containers... **HOW THE HELL** does gluetun it's magic and know which port belongs to which container??! 😳😮

Even everything is working as expected, this question is driving my crazy... 😄

Thank u in advance for clearing.

**Comment by bnhf** [comment]
@Kamikaze01git 

From a networking perspective, whether you're using service:gluetun in a single stack or container:gluetun in separate stacks, only Gluetun has an IP address. All containers referenced through either method are part of the same "localhost", and no two can use the same port number. Gluetun is advertising the port, but only the one container can listen on a given port.

EDIT: And, I suppose the other point worth making, is this isn't really Gluetun magic -- it's Docker magic. :-)

**Reply by Kamikaze01git** [reply]
Thank u @bnhf for your information.

> All containers [...] are part of the same "localhost", and no two can use the same port number. Gluetun is advertising the port, but only the one container can listen on a given port.

okay, I understand that they are all on the same "localhost", but why is (in my example) port 7181 advertised for container 1 and not for container 2?
Is it possible to change them?
What would be the right environments if I want to change the ports for my two containers?

> EDIT: And, I suppose the other point worth making, is this isn't really Gluetun magic -- it's Docker magic. :-)

😄 yeah... thanks for clearing this too... As gluetun is a container in my docker, this stuff is all kind of magic for me 😊 ... Magic I love.

**Reply by bnhf** [reply, marked as answer]
@Kamikaze01git 

What ports Gluetun advertises (exposes) is decided by you when you specify them in the ports section, as you've shown in your original answer. You can map a container's port to a different port by changing the first number in the pairing:

```yml
  ports:
    81:80 # This would change the exposed port to 81, but direct the traffic to 80 internally.
```

However, there can still be only one container listening on port 80. Changing a container's listening port is something that needs to be supported by the application itself through an environment variable. If two containers listen on port 80, and neither can be changed to a different port they can't be used together when using container:gluetun or service:gluetun.

**Reply by Kamikaze01git** [reply]
@bnhf 

> What ports Gluetun advertises (exposes) is decided by you when you specify them in the ports section,...

Yeah - I am aware of port mapping and think I understand that!

> Changing a container's listening port is something that needs to be supported by the application itself

THAT was my missing thing!!! Now I got it sorted 😄👍 thank u very much!!!
I did not set an environment variable for a listing port on both of my containers. Because of that I was wondering why gluetun knows which port is for which container...
But after checking the settings in the application I see, that there is a port assigned 🙏👍
NOW it is clear and I understand the magic 🪄

Thank u very much @bnhf for helping me out of that 😉

Everything clear now - I'm happy again 🫶

**Comment by menghuu** [comment]
If I understand the contents of this [link](https://forums.docker.com/t/diffcult-to-find-documentation-about-how-network-mode-service-service-name-works/137008/2) correctly, this is the function provided by the namespace in the linux kernel. When other docker use network_mode:gluetun, they actually share the same network namespace with gluetun, that is, their ip is exactly the same, and their port range is also shared. Just like two processes on the same machine, they can be directly connected through localhost.
